ancv/web/server.py

```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
+        LOGGER.debug("Adding routes.")
+        self.app.add_routes([web.get("/", self.root)])
+
+        self.app.cleanup_ctx.append(self.app_context)
+
+    def run(self, context: ServerContext) -> None:
+        LOGGER.info("Loaded, starting server...")
+        web.run_app(self.app, host=context.host, port=context.port, path=context.path)
+
+    async def app_context(self, app: web.Application) -> AsyncGenerator[None, None]:
+        """Sets up the application context with required clients.
+
+        Args:
+            app: The app instance to attach our state to.
+        """
+        log = LOGGER.bind(app=app)
+        log.debug("App context initialization starting.")
+
+        log.debug("Starting client session.")
+        session = ClientSession()
+        app["client_session"] = session
+        log.debug("Started client session.")
+
+        log.debug("App context initialization done, yielding.")
+        yield
+
+        log.debug("App context teardown starting.")
+        await session.close()
+        log.debug("App context teardown done.")
+
+    async def fetch_and_render(self, session: ClientSession) -> str:
+        """Fetches the resume from URL and renders it.
+
+        Args:
+            session: The aiohttp client session to use for requests.
+
+        Returns:
+            The rendered template as string.
+        """
+        log = get_logger()
+        stopwatch = Stopwatch()
+        stopwatch(segment="Fetch Resume")
+
+        try:
+            async with session.get(self.destination) as response:
+                if response.status != 200:
+                    raise web.HTTPBadRequest(
+                        reason=f"Failed to fetch resume from {self.destination}"
+                    )
+                content = await response.text()
+                resume_json = json.loads(content)
+
+            with tempfile.NamedTemporaryFile(mode="w", suffix=".json") as tmp:
+                json.dump(resume_json, tmp)
+                tmp.flush()
+                template = Template.from_file(Path(tmp.name))
+                rendered = template.render()
+
+            log.info("Successfully refreshed resume data")
+            self._last_valid_render = rendered
+            return rendered
+
+        except Exception as e:
+            if hasattr(self, "_last_valid_render"):
+                log.warning(
+                    f"Failed to refresh resume, using cached version. Error: {str(e)}"
+                )
+                return self._last_valid_render
+            raise
+
+    async def root(self, request: web.Request) -> web.Response:
+        """The root endpoint, returning the rendered template with periodic refresh.
+
+        Args:
+            request: The incoming web request.
+
+        Returns:
+            The response containing the rendered resume.
+        """
+        log = LOGGER.bind(request=request)
+        session: ClientSession = request.app["client_session"]
+
+        current_time = time.time()
```

Comment:
Similar to timedelta above, let's have this...

- either be a stronger time: https://docs.python.org/3/library/datetime.html#datetime.datetime.now
- or, maybe "semantically better", a monotonic tick: https://docs.python.org/3/library/time.html#time.monotonic

`time.time()` has problems like:

> it can return a lower value than a previous call if the system clock has been set back between the two calls.

which probably messes with the cache.

Reply:
Done, and nice catch, moved to `time.monotonic()`. TIL